File: lib/product/semantic_conformance.ml
type source = { id : int; path : string; provider : Ir.provider }
type dense_node = { id : int; source : int; node : Ir.node }
type dense_edge = { from_ : int; to_ : int; edge : Ir.edge }

let normalized_path path = Util.normalize_slashes path

let sources (report : Report.t) =
  report.graphs
  |> List.map (fun (graph : Ir.t) ->
      (normalized_path graph.source, graph.provider))
  |> List.sort_uniq
       (fun (left_path, left_provider) (right_path, right_provider) ->
         match String.compare left_path right_path with
         | 0 -> Stdlib.compare left_provider right_provider
         | comparison -> comparison)
  |> List.mapi (fun id (path, provider) -> { id; path; provider })

let source_id ?provider sources path =
  let path = normalized_path path in
  sources
  |> List.find_opt (fun (source : source) ->
      source.path = path
      && Option.fold ~none:true
           ~some:(fun provider -> source.provider = provider)
           provider)
  |> Option.map (fun (source : source) -> source.id)
  |> Option.value ~default:0

let position_json (position : Span.position) =
  Json.Object
    [
      ("byte", Json.Int position.byte);
      ("column", Json.Int position.column);
      ("line", Json.Int position.line);
    ]

let span_json sources (span : Span.t) =
  Json.Object
    [
      ("source", Json.Int (source_id sources span.file));
      ("start", position_json span.start);
      ("stop", position_json span.stop);
    ]

let rec normalize_spans sources = function
  | Json.Object fields -> (
      match
        ( List.assoc_opt "file" fields,
          List.assoc_opt "start" fields,
          List.assoc_opt "stop" fields )
      with
      | Some (Json.String file), Some start, Some stop ->
          Json.Object
            [
              ("source", Json.Int (source_id sources file));
              ("start", normalize_spans sources start);
              ("stop", normalize_spans sources stop);
            ]
      | _ ->
          Json.Object
            (List.map
               (fun (name, child) -> (name, normalize_spans sources child))
               fields))
  | Json.Array values -> Json.Array (List.map (normalize_spans sources) values)
  | (Json.Null | Json.Bool _ | Json.Int _ | Json.Int64 _ | Json.String _) as
    value -> value

let compare_node (left : Ir.node) (right : Ir.node) =
  match
    String.compare
      (normalized_path left.span.file)
      (normalized_path right.span.file)
  with
  | 0 -> (
      match Span.compare left.span right.span with
      | 0 -> (
          match Stdlib.compare left.provider right.provider with
          | 0 -> (
              match Stdlib.compare left.kind right.kind with
              | 0 -> (
                  match String.compare left.name right.name with
                  | 0 -> Stdlib.compare left.phase right.phase
                  | comparison -> comparison)
              | comparison -> comparison)
          | comparison -> comparison)
      | comparison -> comparison)
  | comparison -> comparison

let dense_program report sources =
  let program = Program_graph.compose report.Report.graphs in
  let nodes = List.sort compare_node program.nodes in
  let ids = Hashtbl.create (max 1 (List.length nodes)) in
  let nodes =
    List.mapi
      (fun id (node : Ir.node) ->
        Hashtbl.replace ids node.id id;
        {
          id;
          source = source_id ~provider:node.provider sources node.span.file;
          node;
        })
      nodes
  in
  let dense_id id = Hashtbl.find_opt ids id |> Option.value ~default:(-1) in
  let edges =
    program.edges
    |> List.filter_map (fun (edge : Ir.edge) ->
        let from_ = dense_id edge.from_ and to_ = dense_id edge.to_ in
        if from_ < 0 || to_ < 0 || from_ = to_ then None
        else Some { from_; to_; edge })
    |> List.sort_uniq (fun left right ->
        match Int.compare left.from_ right.from_ with
        | 0 -> (
            match Int.compare left.to_ right.to_ with
            | 0 -> (
                match Stdlib.compare left.edge.kind right.edge.kind with
                | 0 -> (
                    match
                      String.compare
                        (Condition.to_string left.edge.condition)
                        (Condition.to_string right.edge.condition)
                    with
                    | 0 ->
                        Option.compare String.compare left.edge.label
                          right.edge.label
                    | comparison -> comparison)
                | comparison -> comparison)
            | comparison -> comparison)
        | comparison -> comparison)
  in
  let entrypoints =
    program.entrypoints |> List.map dense_id
    |> List.filter (( <= ) 0)
    |> List.sort_uniq Int.compare
  in
  (nodes, edges, entrypoints, dense_id)

let source_json (source : source) =
  Json.Object
    [
      ("id", Json.Int source.id);
      ("path", Json.String source.path);
      ("provider", Json.String (Ir.provider_name source.provider));
    ]

let node_json sources (value : dense_node) =
  let node = value.node in
  let fields =
    [
      ("id", Json.Int value.id);
      ("kind", Json.String (Ir.kind_name node.kind));
      ("name", Json.String node.name);
      ("phase", Json.String (Ir.phase_name node.phase));
      ("source", Json.Int value.source);
      ("span", span_json sources node.span);
    ]
  in
  let fields =
    if node.attributes = [] then fields
    else
      ( "attributes",
        Json.Object
          (List.map
             (fun (name, value) ->
               (name, Abstract_value.to_json value |> normalize_spans sources))
             node.attributes) )
      :: fields
  in
  let fields =
    if node.capabilities = [] then fields
    else
      ( "capabilities",
        Json.Array
          (List.map
             (fun value -> Json.String (Ir.capability_name value))
             node.capabilities) )
      :: fields
  in
  let fields =
    if Condition.equal node.condition Condition.true_ then fields
    else ("condition", Condition.to_json node.condition) :: fields
  in
  let fields =
    if node.effects = [] then fields
    else
      ( "effects",
        Json.Array
          (List.map
             (fun value -> Json.String (Ir.effect_name value))
             node.effects) )
      :: fields
  in
  let fields =
    Option.fold ~none:fields
      ~some:(fun unknown -> ("unknown", Unknown.to_json unknown) :: fields)
      node.unknown
  in
  Json.Object fields

let edge_json value =
  let edge = value.edge in
  let fields =
    [
      ("from", Json.Int value.from_);
      ("kind", Json.String (Ir.edge_kind_name edge.kind));
      ("to", Json.Int value.to_);
    ]
  in
  let fields =
    if Condition.equal edge.condition Condition.true_ then fields
    else ("condition", Condition.to_json edge.condition) :: fields
  in
  let fields =
    Option.fold ~none:fields
      ~some:(fun label -> ("label", Json.String label) :: fields)
      edge.label
  in
  Json.Object fields

let fix_json sources (fix : Diagnostic.fix) =
  let fields =
    [
      ("description", Json.String fix.description);
      ("kind", Json.String fix.kind);
    ]
  in
  let fields =
    Option.fold ~none:fields
      ~some:(fun replacement ->
        ("replacement", Json.String replacement) :: fields)
      fix.replacement
  in
  let fields =
    Option.fold ~none:fields
      ~some:(fun span -> ("span", span_json sources span) :: fields)
      fix.span
  in
  Json.Object fields

let diagnostic_json sources dense_id (diagnostic : Diagnostic.t) =
  let fields =
    [
      ( "confidence",
        Json.String (Diagnostic.confidence_name diagnostic.confidence) );
      ("message", Json.String diagnostic.message);
      ("rule_id", Json.String diagnostic.rule_id);
      ("severity", Json.String (Diagnostic.severity_name diagnostic.severity));
      ("span", span_json sources diagnostic.span);
    ]
  in
  let fields =
    if diagnostic.capabilities = [] then fields
    else
      ( "capabilities",
        Json.Array
          (List.map
             (fun value -> Json.String (Ir.capability_name value))
             diagnostic.capabilities) )
      :: fields
  in
  let fields =
    if diagnostic.evidence = [] then fields
    else
      ( "evidence",
        Json.Array
          (List.map (fun value -> Json.String value) diagnostic.evidence) )
      :: fields
  in
  let fields =
    Option.fold ~none:fields
      ~some:(fun fix -> ("fix", fix_json sources fix) :: fields)
      diagnostic.fix
  in
  let fields =
    if diagnostic.trace = [] then fields
    else
      ( "trace",
        Json.Array
          (List.map
             (fun (hop : Diagnostic.trace_hop) ->
               Json.Object
                 [
                   ("label", Json.String hop.label);
                   ("node", Json.Int (dense_id hop.node_id));
                   ("span", span_json sources hop.span);
                 ])
             diagnostic.trace) )
      :: fields
  in
  Json.Object fields

let property_json (property : Property.t) =
  let fields =
    [
      ("explanation", Json.String property.explanation);
      ("id", Json.String property.id);
      ("state", Json.String (Property.state_name property.state));
    ]
  in
  let fields =
    match property.state with
    | Property.Unknown reasons ->
        ("reasons", Json.Array (List.map Unknown.to_json reasons)) :: fields
    | Proved | Violated | Not_applicable -> fields
  in
  let fields =
    Option.fold ~none:fields
      ~some:(fun subject -> ("subject", Json.String subject) :: fields)
      property.subject
  in
  Json.Object fields

let to_canonical_json (report : Report.t) =
  let sources = sources report in
  let nodes, edges, entrypoints, dense_id = dense_program report sources in
  let diagnostics = Report.diagnostics report in
  let properties =
    report.verifications
    |> List.concat_map (fun result -> result.Verifier.properties)
    |> List.sort Property.compare
  in
  let completeness_reasons =
    Util.deduplicate_strings report.provenance.completeness_reasons
  in
  let fields digest =
    [
      ( "completeness",
        Json.Object
          [
            ( "reasons",
              Json.Array
                (List.map (fun value -> Json.String value) completeness_reasons)
            );
            ( "state",
              Json.String
                (if completeness_reasons = [] then "complete" else "incomplete")
            );
          ] );
      ( "diagnostics",
        Json.Array (List.map (diagnostic_json sources dense_id) diagnostics) );
      ("digest", digest);
      ("edges", Json.Array (List.map edge_json edges));
      ( "entrypoints",
        Json.Array (List.map (fun value -> Json.Int value) entrypoints) );
      ( "gate",
        Json.Object
          [
            ("exit_code", Json.Int report.provenance.exit_code);
            ( "result",
              Json.String
                (Report.gate_result_name report.provenance.gate_result) );
          ] );
      ("nodes", Json.Array (List.map (node_json sources) nodes));
      ("properties", Json.Array (List.map property_json properties));
      ("schema", Json.String "semantic-conformance/1");
      ("sources", Json.Array (List.map source_json sources));
    ]
  in
  let digest =
    "sha256:"
    ^ Sha256.digest_string (Json.to_string (Json.Object (fields Json.Null)))
  in
  Json.to_string (Json.Object (fields (Json.String digest))) ^ "\n"

File: src/product/semantic_conformance.rs
use crate::domain::Program;
use crate::foundation::content_digest;
use crate::product::graph_output::{
    EdgesView, GraphDocumentView, GraphKind, NodeIdsView, NodesView, SourcesView,
};
use crate::product::report::{CompletenessView, Gate, GateView, PropertiesView};
use crate::product::serde_views::SpanView;
use crate::verifier::{Diagnostic, Property};
use serde::ser::{Serialize, SerializeMap, SerializeSeq, Serializer};

/// Test-only, implementation-neutral projection used by the OCaml oracle.
#[derive(Clone, Copy)]
pub struct SemanticConformanceView<'a> {
    program: &'a Program,
    diagnostics: &'a [Diagnostic],
    properties: &'a [Property],
    gate: Gate,
    completeness_reasons: &'a [String],
}

impl<'a> SemanticConformanceView<'a> {
    #[must_use]
    pub const fn new(
        program: &'a Program,
        diagnostics: &'a [Diagnostic],
        properties: &'a [Property],
        gate: Gate,
        completeness_reasons: &'a [String],
    ) -> Self {
        Self {
            program,
            diagnostics,
            properties,
            gate,
            completeness_reasons,
        }
    }

    #[must_use]
    /// Computes the semantic conformance document digest.
    ///
    /// # Panics
    ///
    /// Panics only if serialization of the built-in conformance view violates
    /// an internal invariant.
    pub fn digest(self) -> String {
        let bytes = serde_json::to_vec(&Projection {
            view: self,
            digest: None,
        })
        .expect("serializing semantic conformance cannot fail");
        content_digest(bytes)
    }

    #[must_use]
    /// Serializes the test-only semantic conformance document.
    ///
    /// # Panics
    ///
    /// Panics only if serialization of the built-in conformance view violates
    /// an internal invariant.
    pub fn to_canonical_json(self) -> String {
        let mut output =
            serde_json::to_string(&self).expect("serializing semantic conformance cannot fail");
        output.push('\n');
        output
    }
}

struct Projection<'a> {
    view: SemanticConformanceView<'a>,
    digest: Option<&'a str>,
}

fn serialize_document<S: Serializer>(
    view: SemanticConformanceView<'_>,
    digest: Option<&str>,
    serializer: S,
) -> Result<S::Ok, S::Error> {
    let graph = GraphDocumentView::new(GraphKind::All, view.program);
    let mut map = serializer.serialize_map(Some(10))?;
    map.serialize_entry("completeness", &CompletenessView(view.completeness_reasons))?;
    map.serialize_entry("diagnostics", &SemanticDiagnostics(view.diagnostics))?;
    map.serialize_entry("digest", &digest)?;
    map.serialize_entry("edges", &EdgesView(graph))?;
    map.serialize_entry("entrypoints", &NodeIdsView(&view.program.entrypoints))?;
    map.serialize_entry("gate", &GateView(view.gate))?;
    map.serialize_entry("nodes", &NodesView(view.program))?;
    map.serialize_entry("properties", &PropertiesView(view.properties))?;
    map.serialize_entry("schema", "semantic-conformance/1")?;
    map.serialize_entry("sources", &SourcesView(&view.program.sources))?;
    map.end()
}

struct SemanticDiagnostics<'a>(&'a [Diagnostic]);

impl Serialize for SemanticDiagnostics<'_> {
    fn serialize<S: Serializer>(&self, serializer: S) -> Result<S::Ok, S::Error> {
        let mut sequence = serializer.serialize_seq(Some(self.0.len()))?;
        for diagnostic in self.0 {
            sequence.serialize_element(&SemanticDiagnostic(diagnostic))?;
        }
        sequence.end()
    }
}

struct SemanticDiagnostic<'a>(&'a Diagnostic);

impl Serialize for SemanticDiagnostic<'_> {
    fn serialize<S: Serializer>(&self, serializer: S) -> Result<S::Ok, S::Error> {
        let value = self.0;
        let fields = 5
            + usize::from(!value.capabilities.is_empty())
            + usize::from(!value.evidence.is_empty())
            + usize::from(value.fix.is_some())
            + usize::from(!value.trace.is_empty());
        let mut map = serializer.serialize_map(Some(fields))?;
        if !value.capabilities.is_empty() {
            map.serialize_entry(
                "capabilities",
                &value
                    .capabilities
                    .iter()
                    .map(|capability| capability.name())
                    .collect::<Vec<_>>(),
            )?;
        }
        map.serialize_entry("confidence", value.confidence.name())?;
        if !value.evidence.is_empty() {
            map.serialize_entry("evidence", &value.evidence)?;
        }
        if let Some(fix) = &value.fix {
            map.serialize_entry("fix", &SemanticFix(fix))?;
        }
        map.serialize_entry("message", &value.message)?;
        map.serialize_entry("rule_id", &value.rule_id)?;
        map.serialize_entry("severity", value.severity.name())?;
        map.serialize_entry("span", &SpanView::from(value.span))?;
        if !value.trace.is_empty() {
            map.serialize_entry("trace", &SemanticTrace(&value.trace))?;
        }
        map.end()
    }
}

struct SemanticFix<'a>(&'a crate::verifier::Fix);

impl Serialize for SemanticFix<'_> {
    fn serialize<S: Serializer>(&self, serializer: S) -> Result<S::Ok, S::Error> {
        let mut map = serializer.serialize_map(None)?;
        map.serialize_entry("description", &self.0.description)?;
        map.serialize_entry("kind", &self.0.kind)?;
        if let Some(replacement) = &self.0.replacement {
            map.serialize_entry("replacement", replacement)?;
        }
        if let Some(span) = self.0.span {
            map.serialize_entry("span", &SpanView::from(span))?;
        }
        map.end()
    }
}

struct SemanticTrace<'a>(&'a [crate::verifier::TraceHop]);

impl Serialize for SemanticTrace<'_> {
    fn serialize<S: Serializer>(&self, serializer: S) -> Result<S::Ok, S::Error> {
        let mut sequence = serializer.serialize_seq(Some(self.0.len()))?;
        for hop in self.0 {
            sequence.serialize_element(&SemanticHop(hop))?;
        }
        sequence.end()
    }
}

struct SemanticHop<'a>(&'a crate::verifier::TraceHop);

impl Serialize for SemanticHop<'_> {
    fn serialize<S: Serializer>(&self, serializer: S) -> Result<S::Ok, S::Error> {
        let mut map = serializer.serialize_map(Some(3))?;
        map.serialize_entry("label", &self.0.label)?;
        map.serialize_entry("node", &self.0.node_id.0)?;
        map.serialize_entry("span", &SpanView::from(self.0.span))?;
        map.end()
    }
}

impl Serialize for Projection<'_> {
    fn serialize<S: Serializer>(&self, serializer: S) -> Result<S::Ok, S::Error> {
        serialize_document(self.view, self.digest, serializer)
    }
}

impl Serialize for SemanticConformanceView<'_> {
    fn serialize<S: Serializer>(&self, serializer: S) -> Result<S::Ok, S::Error> {
        let digest = self.digest();
        serialize_document(*self, Some(&digest), serializer)
    }
}
